source/index.ts

```diff
@@ -0,0 +1,744 @@
+		}
+	}
+
+	_watch(): void {
+		this._ensureDirectory();
+
+		if (!fs.existsSync(this.path)) {
+			this._write({});
+		}
+
+		fs.watch(this.path, {persistent: false}, debounceFn(() => {
+			// On Linux and Windows, writing to the config file emits a `rename` event, so we skip checking the event type.
+			this.events?.emit('change');
+		}, {wait: 100}));
+	}
+
+	_migrate(migrations: Migrations<T>, versionToMigrate: string): void {
+		let previousMigratedVersion: string = this._get(MIGRATION_KEY, '0.0.0') as string;
+
+		const newerVersions: string[] = Object.keys(migrations)
+			.filter(candidateVersion => this._shouldPerformMigration(candidateVersion, previousMigratedVersion, versionToMigrate));
+
+		let storeBackup = {...this.store};
+
+		for (const version of newerVersions) {
+			try {
+				const migration = migrations[version];
+				migration(this);
+
+				this._set(MIGRATION_KEY, version);
+
+				previousMigratedVersion = version;
+				storeBackup = {...this.store};
+			} catch (error) {
+				this.store = storeBackup;
+
+				throw new Error(
+					`Something went wrong during the migration! Changes applied to the store until this failed migration will be restored. ${error as string}`
+				);
+			}
+		}
+
+		if (this._isVersionInRangeFormat(previousMigratedVersion) || !semver.eq(previousMigratedVersion, versionToMigrate)) {
+			this._set(MIGRATION_KEY, versionToMigrate);
+		}
+	}
+
+	_containsReservedKey(key: string | {[key: string]: unknown}): boolean {
+		if (typeof key === 'object') {
+			const firstKey = Object.keys(key)[0];
+
+			if (firstKey === INTERNAL_KEY) {
+				return true;
+			}
+		}
+
+		if (typeof key !== 'string') {
+			return false;
+		}
+
+		if (this._options.accessPropertiesByDotNotation) {
+			if (key.startsWith(`${INTERNAL_KEY}.`)) {
+				return true;
+			}
+
+			return false;
+		}
+
+		return false;
+	}
+
+	_isVersionInRangeFormat(version: string): boolean {
+		return semver.clean(version) === null;
+	}
+
+	_shouldPerformMigration(candidateVersion: string, previousMigratedVersion: string, versionToMigrate: string): boolean {
+		if (this._isVersionInRangeFormat(candidateVersion)) {
+			if (previousMigratedVersion !== '0.0.0' && semver.satisfies(previousMigratedVersion, candidateVersion)) {
+				return false;
+			}
+
+			return semver.satisfies(versionToMigrate, candidateVersion);
+		}
+
+		if (semver.lte(candidateVersion, previousMigratedVersion)) {
+			return false;
+		}
+
+		if (semver.gt(candidateVersion, versionToMigrate)) {
+			return false;
+		}
+
+		return true;
+	}
+
+	_get(key: string, defaultValue?: unknown): unknown {
+		return dotProp.get(this.store, key, defaultValue);
+	}
+
+	_set(key: string, value?: unknown): void {
+		const {store} = this;
+		dotProp.set(store, key, value);
+
+		this.store = store;
+	}
+
+	/**
+	Get an item.
+
+	@param key - The key of the item to get.
+	@param defaultValue - The default value if the item does not exist.
+	*/
+	get(key: string, defaultValue?: T): T | undefined {
```

Comment:
This should have an overload that makes the return value `T` when `defaultValue` is given.

So the definitions should be:

```ts
get(key: string): T | undefined {
get(key: string, defaultValue: T): T {
```

Reply:
Are you sure about the return type of the second overload?
There's a wide range of inputs to the function where it could return undefined.

one example, assuming the store has: 
`[cool-key]: undefined`

using `get('cool-key', 'this should be the fallback`)` will return `undefined`.

Is also raises the question:
- Do we want undefined, but declared, values in the store to be returned or fallbacked?

Reply:
The store cannot have `undefined`:

>The value must be JSON serializable. Trying to set the type undefined, function, or symbol will result in a TypeError.

My suggestion assumes you also do https://github.com/sindresorhus/conf/pull/96#discussion_r362317785 to all the set/get/reset/etc methods, so they're strongly typed.